### PR TITLE
`paasta metastatus` for Kubernetes

### PR DESCRIPTION
**Pull request by vkhromov**
```
$ PAASTA_SYSTEM_CONFIG_DIR=$HOME/paasta-config USE_API_ENDPOINT=1 paasta metastatus -c kubestage
Cluster: kubestage
No dashboards configured for kubestage!
Master paasta_tools version: 0.81.3
Mesos Status: OK
Marathon Status: OK
Kubernetes Status: OK
Chronos Status: OK
```
```
$ PAASTA_SYSTEM_CONFIG_DIR=$HOME/paasta-config USE_API_ENDPOINT=1 paasta metastatus -c kubestage -v
Cluster: kubestage
No dashboards configured for kubestage!
Master paasta_tools version: 0.81.3
Mesos Status: OK
  Mesos is not configured to run here
Marathon Status: OK
  Marathon is not configured to run here
Kubernetes Status: OK
  Kubernetes deployments:   8
  Pods: running: 30 pending: 1 failed: 0
  CPUs: 0.00 / 54 in use (0.00%)
  Memory: 0.59 / 110.83GB in use (0.53%)
  Disk: 46.49 / 464.94GB in use (10.00%)
  No GPUs found!
  Nodes: active: 6 inactive: 0
Chronos Status: OK
  Chronos is not configured to run here
```
```
$ PAASTA_SYSTEM_CONFIG_DIR=$HOME/paasta-config USE_API_ENDPOINT=1 paasta metastatus -c kubestage -vv
Cluster: kubestage
No dashboards configured for kubestage!
Master paasta_tools version: 0.81.3
Mesos Status: OK
  Mesos is not configured to run here
Marathon Status: OK
  Marathon is not configured to run here
Kubernetes Status: OK
  Kubernetes deployments:   8
  Pods: running: 30 pending: 1 failed: 0
  CPUs: 0.00 / 54 in use (0.00%)
  Memory: 0.59 / 110.83GB in use (0.53%)
  Disk: 46.49 / 464.94GB in use (10.00%)
  No GPUs found!
  Nodes: active: 6 inactive: 0
  Resources Grouped by region
    Region        CPU (used/total)  RAM (used/total)       Disk (used/total)      GPU (used/total)  Agent count
    uswest1-devc  0.00/54 (0.00%)   600.0M/110.8G (0.53%)  46.5G/464.9G (10.00%)  0.00/0 (100.00%)  6
Chronos Status: OK
  Chronos is not configured to run here
```
```
$ PAASTA_SYSTEM_CONFIG_DIR=$HOME/paasta-config USE_API_ENDPOINT=1 paasta metastatus -c kubestage -vvv -a
Cluster: kubestage
No dashboards configured for kubestage!
Master paasta_tools version: 0.81.3
Mesos Status: OK
  Mesos is not configured to run here
Marathon Status: OK
  Marathon is not configured to run here
Kubernetes Status: OK
  Kubernetes deployments:   8
  Pods: running: 30 pending: 1 failed: 0
  CPUs: 0.00 / 54 in use (0.00%)
  Memory: 0.59 / 110.83GB in use (0.53%)
  Disk: 46.49 / 464.94GB in use (10.00%)
  No GPUs found!
  Nodes: active: 6 inactive: 0
  Resources Grouped by region
    Region        CPU (used/total)  RAM (used/total)       Disk (used/total)      GPU (used/total)  Agent count
    uswest1-devc  0.00/54 (0.00%)   600.0M/110.8G (0.53%)  46.5G/464.9G (10.00%)  0.00/0 (100.00%)  6
  No autoscaling resources for Kubernetes
  Per Node Utilization
    Region        Hostname                                    CPU (used/total)  RAM (used/total)      Disk (used/total)    Agent count
    uswest1-devc  10-40-10-205-uswest1adevc.dev.yelpcorp.com  0.00/16 (0.00%)   100.0M/29.4G (0.33%)  5.8G/58.1G (10.00%)  1
    uswest1-devc  10-40-10-244-uswest1adevc.dev.yelpcorp.com  0.00/16 (0.00%)   100.0M/29.4G (0.33%)  5.8G/58.1G (10.00%)  1
    uswest1-devc  10-40-11-189-uswest1adevc.dev.yelpcorp.com  0.00/16 (0.00%)   100.0M/29.4G (0.33%)  5.8G/58.1G (10.00%)  1
    uswest1-devc  kubestage1-uswest1adevc.dev.yelpcorp.com    0.00/2 (0.00%)    100.0M/7.5G (1.30%)   9.7G/96.9G (10.00%)  1
    uswest1-devc  kubestage1-uswest1cdevc.dev.yelpcorp.com    0.00/2 (0.00%)    100.0M/7.5G (1.30%)   9.7G/96.9G (10.00%)  1
    uswest1-devc  kubestage2-uswest1cdevc.dev.yelpcorp.com    0.00/2 (0.00%)    100.0M/7.5G (1.30%)   9.7G/96.9G (10.00%)  1
Chronos Status: OK
  Chronos is not configured to run here
```

### CODE REVIEW

File: paasta_tools/metrics/metastatus_lib.py [hunk 1]
```diff
@@ -267,7 +427,23 @@ def assert_tasks_running(metrics):
...
+def assert_kube_pods_running(
+    kube_client: KubeClient,
+) -> HealthCheckResult:
+    statuses = [get_pod_status(pod) for pod in get_all_pods(kube_client)]
+    running = statuses.count(PodStatus.RUNNING)
+    pending = statuses.count(PodStatus.PENDING)
+    failed = statuses.count(PodStatus.FAILED)
+    return HealthCheckResult(
+        message=f'Pods: running: {running} pending: {pending} failed: {failed}',
+        healthy=True,
```
Comment (mattmb) [inline paasta_tools/metrics/metastatus_lib.py]: Should we maybe only return true if at least 1 pod is running? Do we do something similar for marathon apps iirc?

File: paasta_tools/metrics/metastatus_lib.py [hunk 2]
```diff
@@ -321,15 +500,30 @@ def assert_frameworks_exist(state, expected):
...
+
+
+def assert_nodes_health(
+    nodes_health_status: Tuple[int, int],
+) -> HealthCheckResult:
+    active, inactive = nodes_health_status
     return HealthCheckResult(
-        message="Slaves: active: %d inactive: %d" % (active, inactive),
+        message="Nodes: active: %d inactive: %d" % (active, inactive),
         healthy=True,
```
Comment (mattmb) [inline paasta_tools/metrics/metastatus_lib.py]: As above

File: paasta_tools/metrics/metastatus_lib.py [hunk 3]
```diff
@@ -532,7 +835,40 @@ def get_resource_utilization_by_grouping(
...
 
 
-def resource_utillizations_from_resource_info(total, free):
+def get_resource_utilization_by_grouping_kube(
+    grouping_func: _GroupingFunctionT,
+    kube_client: KubeClient,
+    # filters: Sequence[_SlaveFilterFunctionT]=[],
+    # sort_func=None,
+) -> Mapping[_KeyFuncRetT, ResourceUtilizationDict]:
+    """ Given a function used to group slaves and mesos state, calculate
```
Comment (mattmb) [inline paasta_tools/metrics/metastatus_lib.py]: doc string needs a fixup

File: paasta_tools/metrics/metastatus_lib.py [hunk 4]
```diff
@@ -532,7 +835,40 @@ def get_resource_utilization_by_grouping(
     }
 
 
-def resource_utillizations_from_resource_info(total, free):
+def get_resource_utilization_by_grouping_kube(
+    grouping_func: _GroupingFunctionT,
+    kube_client: KubeClient,
+    # filters: Sequence[_SlaveFilterFunctionT]=[],
```
Comment (mattmb) [inline paasta_tools/metrics/metastatus_lib.py]: ?

File: paasta_tools/paasta_metastatus.py [hunk 5]
```diff
@@ -345,13 +447,49 @@ def print_output(argv: Optional[List[str]]=None) -> None:
             for line in format_table(all_rows):
                 print_with_indent(line, 4)
     metastatus_lib.print_results_for_healthchecks(marathon_summary, marathon_ok, marathon_results, args.verbose)
+    metastatus_lib.print_results_for_healthchecks(kube_summary, kube_ok, kube_results, args.verbose)
+    if args.verbose > 1 and kube_enabled:
```
Comment (mattmb) [inline paasta_tools/paasta_metastatus.py]: I hate how big this function is already :-(

But really don't feel obliged to tackle that here

File: paasta_tools/paasta_metastatus.py [hunk 6]
```diff
@@ -345,13 +447,49 @@ def print_output(argv: Optional[List[str]]=None) -> None:
...
+        if args.autoscaling_info:
+            print_with_indent("No autoscaling resources for Kubernetes", 2)
+
+        if args.verbose >= 3:
+            print_with_indent('Per Node Utilization', 2)
+            cluster = system_paasta_config.get_cluster()
+            service_instance_stats = get_service_instance_stats(args.service, args.instance, cluster)
+            if service_instance_stats:
+                print_with_indent('Service-Instance stats:' + str(service_instance_stats), 2)
+            # print info about slaves here. Note that we don't make modifications to
```
Comment (mattmb) [inline paasta_tools/paasta_metastatus.py]: fixup the comment

File: paasta_tools/paasta_metastatus.py [hunk 7]
```diff
@@ -345,13 +447,49 @@ def print_output(argv: Optional[List[str]]=None) -> None:
...
+            # the healthy_exit variable here, because we don't care about a single slave
+            # having high usage.
+            all_rows, _ = utilization_table_by_grouping_from_kube(
+                groupings=args.groupings + ["hostname"],
+                threshold=args.threshold,
+                kube_client=kube_client,
+                service_instance_stats=service_instance_stats,
+            )
+            # The last column from utilization_table_by_grouping_from_mesos_state is "Agent count", which will always be
```
Comment (mattmb) [inline paasta_tools/paasta_metastatus.py]: And here

File: paasta_tools/metrics/metastatus_lib.py [hunk 8]
```diff
@@ -462,7 +700,66 @@ def calculate_resource_utilization_for_slaves(slaves: List, tasks: List) -> Reso
...
+_IEC_NUMBER_SUFFIXES = {
+    'Ki': 1024,
+    'Mi': 1024**2,
+    'Gi': 1024**3,
+    'Ti': 1024**4,
+    'Pi': 1024**5,
+}
+
+
+def suffixed_number_value(s: str) -> int:
```
[the remark this answers is not in the document]
Comment (solarkennedy) [reply paasta_tools/metrics/metastatus_lib.py]: We have `humanize`, but it goes in the other direction.

File: paasta_tools/paasta_metastatus.py [hunk 9]
```diff
@@ -262,33 +313,68 @@ def get_service_instance_stats(
...
+            message='Mesos is not configured to run here',
+            healthy=True,
+        )]
+
+    if kube_enabled:
+        kube_client = KubeClient()
+        kube_results = _run_kube_checks(kube_client)
+    else:
+        kube_results = [metastatus_lib.HealthCheckResult(
+            message='Kubernetesis not configured to run here',
```
Comment (solarkennedy) [inline paasta_tools/paasta_metastatus.py]: typo

**Comment by solarkennedy** [comment]
Even by looking at only the second commit, this review is very large and I don't feel like I can meaningfully add commentary. 

But it does look like it works!

**Pull request merged**